README.md
# php-pubsub-google-cloud

A Google Cloud adapter for the [php-pubsub](https://github.com/Superbalist/php-pubsub) package.

[![Author](http://img.shields.io/badge/author-@superbalist-blue.svg?style=flat-square)](https://twitter.com/superbalist)
[![Build Status](https://img.shields.io/travis/Superbalist/php-pubsub-google-cloud/master.svg?style=flat-square)](https://travis-ci.org/Superbalist/php-pubsub-google-cloud)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Packagist Version](https://img.shields.io/packagist/v/superbalist/php-pubsub-google-cloud.svg?style=flat-square)](https://packagist.org/packages/superbalist/php-pubsub-google-cloud)
[![Total Downloads](https://img.shields.io/packagist/dt/superbalist/php-pubsub-google-cloud.svg?style=flat-square)](https://packagist.org/packages/superbalist/php-pubsub-google-cloud)


## Installation

```bash
composer require superbalist/php-pubsub-google-cloud
```

## Usage

```php
putenv('GOOGLE_APPLICATION_CREDENTIALS=' . __DIR__ . '/../your-gcloud-key.json');

$client = new \Google\Cloud\PubSub\PubSubClient([
    'projectId' => 'your-project-id-here',
]);

$adapter = new \Superbalist\PubSub\GoogleCloud\GoogleCloudPubSubAdapter($client);


// disable auto topic & subscription creation
$adapter->setAutoCreateTopics(false); // this is true by default
$adapter->setAutoCreateSubscriptions(false); // this is true by default

// set a unique client identifier for the subscriber
$adapter->setClientIdentifier('search_service');

// consume messages
// note: this is a blocking call
$adapter->subscribe('my_channel', function ($message) {
    var_dump($message);
});

// publish messages
$adapter->publish('my_channel', 'HELLO WORLD');
$adapter->publish('my_channel', json_encode(['hello' => 'world']));
$adapter->publish('my_channel', 1);
$adapter->publish('my_channel', false);
```

## gRPC Support

Google Cloud PHP v0.12.0 added support for communication over the gRPC protocol.

> gRPC is great for high-performance, low-latency applications, and is highly recommended in cases where performance and latency are concerns.

The library will automatically choose gRPC over REST if all dependencies are installed.
* [gRPC PECL extension](https://pecl.php.net/package/gRPC)
* [google/proto-client-php composer package](https://github.com/googleapis/gax-php)
* [googleapis/proto-client-php composer package](https://github.com/googleapis/proto-client-php)

```bash
pecl install grpc

composer require google/gax
composer require google/proto-client
```

## Background Batch Message Support

Google Cloud v0.33.0 added support for queueing messages and publishing in the background.  This is available in
version 5+ of this package which requires a min version of google/cloud ^0.33.0.

You can enable background batch messaging by setting `$backgroundBatching` to `true` when constructing the
`GoogleCloudPubSubAdapter` or by calling `setBackgroundBatching(true)` on an existing adapter.

If the [semaphore](http://php.net/manual/en/book.sem.php) and [pcntl](http://php.net/manual/en/book.pcntl.php) PHP extensions are
enabled AND the `IS_BATCH_DAEMON_RUNNING` ENV var is set to `true`, the library will queue messages to be published by
the [Batch Daemon](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/src/Core/Batch/BatchDaemon.php).
The Batch Daemon needs to be manually run as a long-lived background process.

For all other cases, messages will be queued in memory and will be published before the script terminates using a
vendor registered shutdown handler.

**Please Note**

This is marked by google/cloud as an experimental feature & may change before release in backwards-incompatible ways.

## Examples

The library comes with [examples](examples) for the adapter and a [Dockerfile](Dockerfile) for
running the example scripts.

Run `make up`.

You will start at a `bash` prompt in the `/opt/php-pubsub` directory.

If you need another shell to publish a message to a blocking consumer, you can run `docker-compose run php-pubsub-google-cloud /bin/bash`

To run the examples:
```bash
$ php examples/GoogleCloudConsumerExample.php
$ php examples/GoogleCloudPublishExample.php (in a separate shell)
```


changelog.md
# Changelog

## 5.2.0 - 2019-03-19

* Add ability to use returnImmediately flag when pulling messages 

## 5.1.0 - 2019-02-28

* Bump up google/cloud requirement to ^0.95.0

## 5.0.1 - 2018-07-27

* Allow setting of maximum number of messages to pull option

## 5.0.0 - 2017-07-25

* Add support for using Google Cloud batch requests (aka background daemon)

## 4.0.1 - 2017-07-21

* Allow for google/cloud ^0.29.0|^0.30.0|^0.31.0|^0.32.0|^0.33.0|^0.34.0|^0.35.0

## 4.0.0 - 2017-05-16

* Allow for google/cloud ^0.26.0|^0.27.0|^0.28.0
* Bump up to superbalist/php-pubsub ^2.0
* Add new publishBatch method to GoogleCloudPubSubAdapter

## 3.0.1 - 2017-04-03

* Fix to gRPC timeouts
* Allow for google/cloud ^0.21.0|^0.22.0|^0.23.0|^0.24.0|^0.25.0

## 3.0.0 - 2017-01-03

* Bump up google/cloud requirement to ^0.11.0|^0.12.0|^0.13.0|^0.20.0

## 2.0.2 - 2017-01-03

* Allow for google/cloud ^0.10.0

## 2.0.1 - 2016-10-05

* Fix to subscriber bug - client identifier needs to be unique across topics.

## 2.0.0 - 2016-10-05

* Add new `$clientIdentifier` & `setClientIdentifier()` functionality to allow subscribers to use the same, or unique identifiers.

## 1.0.2 - 2016-09-26

* Allow for google/cloud ^0.8.0 and ^0.9.0

## 1.0.1 - 2016-09-15

* Ack messages individually after callable returns successfully
* Add functionality to enable/disable auto topic & subscription creation

## 1.0.0 - 2016-09-05

* Initial release


tests/GoogleCloudPubSubAdapterTest.php
<?php

namespace Tests;

use Google\Cloud\PubSub\BatchPublisher;
use Google\Cloud\PubSub\Message;
use Google\Cloud\PubSub\PubSubClient;
use Google\Cloud\PubSub\Subscription;
use Google\Cloud\PubSub\Topic;
use Mockery;
use PHPUnit\Framework\TestCase;
use Superbalist\PubSub\GoogleCloud\GoogleCloudPubSubAdapter;

class GoogleCloudPubSubAdapterTest extends TestCase
{
    public function testGetClient()
    {
        $client = Mockery::mock(PubSubClient::class);
        $adapter = new GoogleCloudPubSubAdapter($client);
        $this->assertSame($client, $adapter->getClient());
    }

    public function testGetSetClientIdentifier()
    {
        $client = Mockery::mock(PubSubClient::class);
        $adapter = new GoogleCloudPubSubAdapter($client);
        $this->assertNull($adapter->getClientIdentifier());

        $adapter->setClientIdentifier('my_identifier');
        $this->assertEquals('my_identifier', $adapter->getClientIdentifier());
    }

    public function testGetSetAutoCreateTopics()
    {
        $client = Mockery::mock(PubSubClient::class);
        $adapter = new GoogleCloudPubSubAdapter($client);
        $this->assertTrue($adapter->areTopicsAutoCreated());

        $adapter->setAutoCreateTopics(false);
        $this->assertFalse($adapter->areTopicsAutoCreated());
    }

    public function testGetSetAutoCreateSubscriptions()
    {
        $client = Mockery::mock(PubSubClient::class);
        $adapter = new GoogleCloudPubSubAdapter($client);
        $this->assertTrue($adapter->areSubscriptionsAutoCreated());

        $adapter->setAutoCreateSubscriptions(false);
        $this->assertFalse($adapter->areSubscriptionsAutoCreated());
    }

    public function testGetSetBackgroundBatching()
    {
        $client = Mockery::mock(PubSubClient::class);
        $adapter = new GoogleCloudPubSubAdapter($client);
        $this->assertFalse($adapter->isBackgroundBatchingEnabled());

        $adapter->setBackgroundBatching(true);
        $this->assertTrue($adapter->isBackgroundBatchingEnabled());

        $adapter = new GoogleCloudPubSubAdapter($client, null, true, true, true);
        $this->assertTrue($adapter->isBackgroundBatchingEnabled());
    }

    public function testGetSetReturnImmediately()
    {
        $client = Mockery::mock(PubSubClient::class);
        $adapter = new GoogleCloudPubSubAdapter($client);
        $this->assertFalse($adapter->getReturnImmediately());

        $adapter->setReturnImmediately(true);
        $this->assertTrue($adapter->getReturnImmediately());
    }

    public function testGetSetReturnImmediatelyPause()
    {
        $client = Mockery::mock(PubSubClient::class);
        $adapter = new GoogleCloudPubSubAdapter($client);
        $this->assertEquals(500000, $adapter->getReturnImmediatelyPause());

        $adapter->setReturnImmediatelyPause(1000000);
        $this->assertEquals(1000000, $adapter->getReturnImmediatelyPause());
    }

    public function testPublishWhenTopicMustBeCreated()
    {
        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(false);
        $topic->shouldReceive('create')
            ->once();
        $topic->shouldReceive('publish')
            ->with([
                'data' => '{"hello":"world"}',
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);

        $adapter->publish('channel_name', ['hello' => 'world']);
    }

    public function testPublishWhenTopicMustBeCreatedAndBackgroundBatchingIsEnabled()
    {
        $batchPublisher = Mockery::mock(BatchPublisher::class);

        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(false);
        $topic->shouldReceive('create')
            ->once();

        $topic->shouldReceive('batchPublisher')
            ->once()
            ->andReturn($batchPublisher);

        $batchPublisher->shouldReceive('publish')
            ->with([
                'data' => '{"hello":"world"}',
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);
        $adapter->setBackgroundBatching(true);

        $adapter->publish('channel_name', ['hello' => 'world']);
    }

    public function testPublishWhenTopicExists()
    {
        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('publish')
            ->with([
                'data' => '{"hello":"world"}',
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);

        $adapter->publish('channel_name', ['hello' => 'world']);
    }

    public function testPublishWhenTopicExistsAndBackgroundBatchingIsEnabled()
    {
        $batchPublisher = Mockery::mock(BatchPublisher::class);

        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('batchPublisher')
            ->once()
            ->andReturn($batchPublisher);

        $batchPublisher->shouldReceive('publish')
            ->with([
                'data' => '{"hello":"world"}',
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);
        $adapter->setBackgroundBatching(true);

        $adapter->publish('channel_name', ['hello' => 'world']);
    }

    public function testPublishWhenAutoTopicCreationIsDisabled()
    {
        $topic = Mockery::mock(Topic::class);
        $topic->shouldNotHaveReceived('exists');
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('publish')
            ->with([
                'data' => '{"hello":"world"}',
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client, null, false);

        $adapter->publish('channel_name', ['hello' => 'world']);
    }

    public function testPublishWhenAutoTopicCreationIsDisabledAndBackgroundBatchingIsEnabled()
    {
        $batchPublisher = Mockery::mock(BatchPublisher::class);

        $topic = Mockery::mock(Topic::class);
        $topic->shouldNotHaveReceived('exists');
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('batchPublisher')
            ->once()
            ->andReturn($batchPublisher);

        $batchPublisher->shouldReceive('publish')
            ->with([
                'data' => '{"hello":"world"}',
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client, null, false);
        $adapter->setBackgroundBatching(true);

        $adapter->publish('channel_name', ['hello' => 'world']);
    }

    public function testPublishBatch()
    {
        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldReceive('publishBatch')
            ->with([
                ['data' => '{"hello":"world"}'],
                ['data' => '"booo!"'],
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);

        $messages = [
            ['hello' => 'world'],
            'booo!',
        ];
        $adapter->publishBatch('channel_name', $messages);
    }

    public function testPublishBatchWhenBackgroundBatchingIsEnabled()
    {
        $batchPublisher = Mockery::mock(BatchPublisher::class);

        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldReceive('batchPublisher')
            ->once()
            ->andReturn($batchPublisher);

        $batchPublisher->shouldReceive('publish')
            ->with([
                'data' => '{"hello":"world"}',
            ])
            ->once();
        $batchPublisher->shouldReceive('publish')
            ->with([
                'data' => '"booo!"',
            ])
            ->once();

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);
        $adapter->setBackgroundBatching(true);

        $messages = [
            ['hello' => 'world'],
            'booo!',
        ];
        $adapter->publishBatch('channel_name', $messages);
    }

    public function testSubscribeWhenSubscriptionMustBeCreated()
    {
        $message1 = new Message(['data' => '{"hello":"world"}'], ['ackId' => 1]);
        $message2 = new Message(['data' => '"this is a string"'], ['ackId' => 2]);
        $message3 = new Message(['data' => '"unsubscribe"'], ['ackId' => 3]);

        $messageBatch1 = [
            $message1,
            $message2,
        ];
        $messageBatch2 = [
            $message3,
        ];

        $subscription = Mockery::mock(Subscription::class);
        $subscription->shouldReceive('exists')
            ->once()
            ->andReturn(false);
        $subscription->shouldReceive('create')
            ->once();
        $subscription->shouldReceive('pull')
            ->with([
                'grpcOptions' => [
                    'timeoutMillis' => null,
                ],
                'maxMessages' => 1000,
                'returnImmediately' => false
            ])
            ->once()
            ->andReturn($messageBatch1);

        $subscription->shouldReceive('acknowledge')
            ->with($message1)
            ->once();
        $subscription->shouldReceive('acknowledge')
            ->with($message2)
            ->once();
        $subscription->shouldReceive('pull')
            ->with([
                'grpcOptions' => [
                    'timeoutMillis' => null,
                ],
                'maxMessages' => 1000,
                'returnImmediately' => false
            ])
            ->once()
            ->andReturn($messageBatch2);
        $subscription->shouldReceive('acknowledge')
            ->with($message3)
            ->once();

        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('subscription')
            ->with('default.channel_name')
            ->once()
            ->andReturn($subscription);

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);

        $handler1 = Mockery::mock(\stdClass::class);
        $handler1->shouldReceive('handle')
            ->with(['hello' => 'world'])
            ->once();
        $handler1->shouldReceive('handle')
            ->with('this is a string')
            ->once();

        $adapter->subscribe('channel_name', [$handler1, 'handle']);
    }

    public function testSubscribeWhenSubscriptionExists()
    {
        $message1 = new Message(['data' => '{"hello":"world"}'], ['ackId' => 1]);
        $message2 = new Message(['data' => '"this is a string"'], ['ackId' => 2]);
        $message3 = new Message(['data' => '"unsubscribe"'], ['ackId' => 3]);

        $messageBatch1 = [
            $message1,
            $message2,
        ];
        $messageBatch2 = [
            $message3,
        ];

        $subscription = Mockery::mock(Subscription::class);
        $subscription->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $subscription->shouldNotHaveReceived('create');
        $subscription->shouldReceive('pull')
            ->with([
                'grpcOptions' => [
                    'timeoutMillis' => null,
                ],
                'maxMessages' => 1000,
                'returnImmediately' => false
            ])
            ->once()
            ->andReturn($messageBatch1);
        $subscription->shouldReceive('acknowledge')
            ->with($message1)
            ->once();
        $subscription->shouldReceive('acknowledge')
            ->with($message2)
            ->once();
        $subscription->shouldReceive('pull')
            ->with([
                'grpcOptions' => [
                    'timeoutMillis' => null,
                ],
                'maxMessages' => 1000,
                'returnImmediately' => false
            ])
            ->once()
            ->andReturn($messageBatch2);
        $subscription->shouldReceive('acknowledge')
            ->with($message3)
            ->once();

        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('subscription')
            ->with('default.channel_name')
            ->once()
            ->andReturn($subscription);

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client);

        $handler1 = Mockery::mock(\stdClass::class);
        $handler1->shouldReceive('handle')
            ->with(['hello' => 'world'])
            ->once();
        $handler1->shouldReceive('handle')
            ->with('this is a string')
            ->once();

        $adapter->subscribe('channel_name', [$handler1, 'handle']);
    }

    public function testSubscribeWhenAutoTopicCreationIsDisabled()
    {
        $message1 = new Message(['data' => '{"hello":"world"}'], ['ackId' => 1]);
        $message2 = new Message(['data' => '"this is a string"'], ['ackId' => 2]);
        $message3 = new Message(['data' => '"unsubscribe"'], ['ackId' => 3]);

        $messageBatch1 = [
            $message1,
            $message2,
        ];
        $messageBatch2 = [
            $message3,
        ];

        $subscription = Mockery::mock(Subscription::class);
        $subscription->shouldNotHaveReceived('exists');
        $subscription->shouldNotHaveReceived('create');
        $subscription->shouldReceive('pull')
            ->with([
                'grpcOptions' => [
                    'timeoutMillis' => null,
                ],
                'maxMessages' => 1000,
                'returnImmediately' => false
            ])
            ->once()
            ->andReturn($messageBatch1);
        $subscription->shouldReceive('acknowledge')
            ->with($message1)
            ->once();
        $subscription->shouldReceive('acknowledge')
            ->with($message2)
            ->once();
        $subscription->shouldReceive('pull')
            ->with([
                'grpcOptions' => [
                    'timeoutMillis' => null,
                ],
                'maxMessages' => 1000,
                'returnImmediately' => false
            ])
            ->once()
            ->andReturn($messageBatch2);
        $subscription->shouldReceive('acknowledge')
            ->with($message3)
            ->once();

        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('subscription')
            ->with('default.channel_name')
            ->once()
            ->andReturn($subscription);

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $adapter = new GoogleCloudPubSubAdapter($client, null, true, false);

        $handler1 = Mockery::mock(\stdClass::class);
        $handler1->shouldReceive('handle')
            ->with(['hello' => 'world'])
            ->once();
        $handler1->shouldReceive('handle')
            ->with('this is a string')
            ->once();

        $adapter->subscribe('channel_name', [$handler1, 'handle']);
    }

    public function testSubscribeWhenReturnImmediatelyIsEnabled()
    {
        $message1 = new Message(['data' => '{"hello":"world"}'], ['ackId' => 1]);
        $message2 = new Message(['data' => '"this is a string"'], ['ackId' => 2]);
        $message3 = new Message(['data' => '"unsubscribe"'], ['ackId' => 3]);

        $messageBatch1 = [
            $message1,
            $message2,
        ];

        $messageBatch2 = [
            $message3,
        ];

        $subscription = Mockery::mock(Subscription::class);
        $subscription->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $subscription->shouldNotHaveReceived('create');

        $expectedPullOptions = [
            'grpcOptions' => [
                'timeoutMillis' => null,
            ],
            'maxMessages' => 1000,
            'returnImmediately' => true
        ];

        $subscription->shouldReceive('pull')
            ->with($expectedPullOptions)
            ->once()
            ->andReturn($messageBatch1);
        $subscription->shouldReceive('acknowledge')
            ->with($message1)
            ->once();
        $subscription->shouldReceive('acknowledge')
            ->with($message2)
            ->once();

        $subscription->shouldReceive('pull')
            ->with($expectedPullOptions)
            ->once()
            ->andReturn($messageBatch2);
        $subscription->shouldReceive('acknowledge')
            ->with($message3)
            ->once();

        $topic = Mockery::mock(Topic::class);
        $topic->shouldReceive('exists')
            ->once()
            ->andReturn(true);
        $topic->shouldNotHaveReceived('create');
        $topic->shouldReceive('subscription')
            ->with('default.channel_name')
            ->once()
            ->andReturn($subscription);

        $client = Mockery::mock(PubSubClient::class);
        $client->shouldReceive('topic')
            ->with('channel_name')
            ->once()
            ->andReturn($topic);

        $handler1 = Mockery::mock(\stdClass::class);
        $handler1->shouldReceive('handle')
            ->with(['hello' => 'world'])
            ->once();
        $handler1->shouldReceive('handle')
            ->with('this is a string')
            ->once();

        $adapter = new GoogleCloudPubSubAdapter($client);
        $adapter->setReturnImmediately(true);
        $adapter->subscribe('channel_name', [$handler1, 'handle']);
    }
}


src/GoogleCloudPubSubAdapter.php
<?php

namespace Superbalist\PubSub\GoogleCloud;

use Google\Cloud\PubSub\Message;
use Google\Cloud\PubSub\PubSubClient;
use Google\Cloud\PubSub\Subscription;
use Superbalist\PubSub\PubSubAdapterInterface;
use Superbalist\PubSub\Utils;

class GoogleCloudPubSubAdapter implements PubSubAdapterInterface
{
    /**
     * @var PubSubClient
     */
    protected $client;

    /**
     * @var string
     */
    protected $clientIdentifier;

    /**
     * @var bool
     */
    protected $autoCreateTopics;

    /**
     * @var bool
     */
    protected $autoCreateSubscriptions;

    /**
     * @var bool
     */
    protected $backgroundBatching;

    /**
     * @var int
     */
    protected $maxMessages;

    /**
     * @var bool
     */
    protected $returnImmediately;

    /**
     * @var int
     */
    protected $returnImmediatelyPause;

    /**
     * @param PubSubClient $client
     * @param string $clientIdentifier
     * @param bool $autoCreateTopics
     * @param bool $autoCreateSubscriptions
     * @param bool $backgroundBatching
     * @param int $maxMessages
     */
    public function __construct(
        PubSubClient $client,
        $clientIdentifier = null,
        $autoCreateTopics = true,
        $autoCreateSubscriptions = true,
        $backgroundBatching = false,
        $maxMessages = 1000,
        $returnImmediately = false,
        $returnImmediatelyPause = 500000
    ) {
        $this->client = $client;
        $this->clientIdentifier = $clientIdentifier;
        $this->autoCreateTopics = $autoCreateTopics;
        $this->autoCreateSubscriptions = $autoCreateSubscriptions;
        $this->backgroundBatching = $backgroundBatching;
        $this->maxMessages = $maxMessages;
        $this->returnImmediately = $returnImmediately;
        $this->returnImmediatelyPause = (int) $returnImmediatelyPause;
    }

    /**
     * Return the Google PubSubClient.
     *
     * @return PubSubClient
     */
    public function getClient()
    {
        return $this->client;
    }

    /**
     * Set the unique client identifier.
     *
     * The client identifier is used when creating a subscription to a topic.
     *
     * A topic can have multiple subscribers connected.
     * If all subscribers use the same client identifier, the messages will load-balance across them.
     * If all subscribers have different client identifiers, the messages will be dispatched all of them.
     *
     * @param string $clientIdentifier
     */
    public function setClientIdentifier($clientIdentifier)
    {
        $this->clientIdentifier = $clientIdentifier;
    }

    /**
     * Return the unique client identifier.
     *
     * @return string
     */
    public function getClientIdentifier()
    {
        return $this->clientIdentifier;
    }

    /**
     * Set whether or not topics will be auto created.
     *
     * @param bool $autoCreateTopics
     */
    public function setAutoCreateTopics($autoCreateTopics)
    {
        $this->autoCreateTopics = $autoCreateTopics;
    }

    /**
     * Check whether or not topics will be auto created.
     *
     * @return bool
     */
    public function areTopicsAutoCreated()
    {
        return $this->autoCreateTopics;
    }

    /**
     * Set whether or not subscriptions will be auto created.
     *
     * @param bool $autoCreateSubscriptions
     */
    public function setAutoCreateSubscriptions($autoCreateSubscriptions)
    {
        $this->autoCreateSubscriptions = $autoCreateSubscriptions;
    }

    /**
     * Check whether or not subscriptions will be auto created.
     *
     * @return bool
     */
    public function areSubscriptionsAutoCreated()
    {
        return $this->autoCreateSubscriptions;
    }

    /**
     * Set if a pull should return immediately if there are no messages
     * @param bool $returnImmediately
     */
    public function setReturnImmediately($returnImmediately) {
        $this->returnImmediately = $returnImmediately;
    }

    /**
     * Return the return immediately configuration
     * @return bool
     */
    public function getReturnImmediately() {
        return $this->returnImmediately;
    }

    /**
     * Set the amount of time to pause between attempts to pull messages if return immediately is enabled.
     * Value is in microseconds
     *
     * @param int $returnImmediatelyPause
     */
    public function setReturnImmediatelyPause($returnImmediatelyPause) {
        $this->returnImmediatelyPause = (int) $returnImmediatelyPause;
    }

    /**
     * Return the return immediately pause configuration
     * @return int
     */
    public function getReturnImmediatelyPause() {
        return $this->returnImmediatelyPause;
    }

    /**
     * Set whether or not background batching is enabled.
     *
     * This is available from Google Cloud 0.33+ - https://github.com/GoogleCloudPlatform/google-cloud-php/releases/tag/v0.33.0
     *
     * If the http://php.net/manual/en/book.sem.php and http://php.net/manual/en/book.pcntl.php extensions are enabled
     * AND the IS_BATCH_DAEMON_RUNNING ENV var is set to true, the library will queue messages to be published by the
     * Batch Daemon (https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/src/Core/Batch/BatchDaemon.php)
     *
     * For all other cases, messages will be queued in memory and will be published before the script terminates using
     * a vendor registered shutdown handler.
     *
     * @param bool $backgroundBatching
     */
    public function setBackgroundBatching($backgroundBatching)
    {
        $this->backgroundBatching = $backgroundBatching;
    }

    /**
     * Check whether or not background batching is enabled.
     *
     * @return bool
     */
    public function isBackgroundBatchingEnabled()
    {
        return $this->backgroundBatching;
    }

    /**
     * Max messages to pull at a time.
     * https://googlecloudplatform.github.io/google-cloud-php/#/docs/google-cloud/v0.35.0/pubsub/subscription?method=pull
     *
     * @param int $maxMessages
     */
    public function setMaxMessages($maxMessages)
    {
        $this->maxMessages = $maxMessages;
    }

    /**
     * Subscribe a handler to a channel.
     *
     * @param string $channel
     * @param callable $handler
     */
    public function subscribe($channel, callable $handler)
    {
        $subscription = $this->getSubscriptionForChannel($channel);

        $isSubscriptionLoopActive = true;
        $isPauseEnabled = $this->returnImmediately && ($this->returnImmediatelyPause > 0);

        while ($isSubscriptionLoopActive) {
            $messages = $subscription->pull([
                'grpcOptions' => [
                    'timeoutMillis' => null,
                ],
                'maxMessages' => $this->maxMessages,
                'returnImmediately' => $this->returnImmediately,
            ]);
            if ($isPauseEnabled && empty($messages)) {
                usleep($this->returnImmediatelyPause);
                continue;
            }
            foreach ($messages as $message) {
                /** @var Message $message */
                $payload = Utils::unserializeMessagePayload($message->data());

                if ($payload === 'unsubscribe') {
                    $isSubscriptionLoopActive = false;
                } else {
                    call_user_func($handler, $payload);
                }

                $subscription->acknowledge($message);
            }
        }
    }

    /**
     * Publish a message to a channel.
     *
     * @param string $channel
     * @param mixed $message
     */
    public function publish($channel, $message)
    {
        $topic = $this->getTopicForChannel($channel);
        $payload = Utils::serializeMessage($message);

        if ($this->backgroundBatching) {
            $topic->batchPublisher()->publish(['data' => $payload]);
        } else {
            $topic->publish(['data' => $payload]);
        }
    }

    /**
     * Publish multiple messages to a channel.
     *
     * @param string $channel
     * @param array $messages
     */
    public function publishBatch($channel, array $messages)
    {
        $topic = $this->getTopicForChannel($channel);
        $messages = array_map(function ($message) {
            return ['data' => Utils::serializeMessage($message)];
        }, $messages);

        if ($this->backgroundBatching) {
            $batchPublisher = $topic->batchPublisher();
            foreach ($messages as $message) {
                $batchPublisher->publish($message);
            }
        } else {
            $topic->publishBatch($messages);
        }
    }

    /**
     * Return a `Topic` instance from a channel name.
     *
     * If the topic doesn't exist, the topic is first created.
     *
     * @param string $channel
     *
     * @return \Google\Cloud\PubSub\Topic
     */
    protected function getTopicForChannel($channel)
    {
        $topic = $this->client->topic($channel);
        if ($this->autoCreateTopics && !$topic->exists()) {
            $topic->create();
        }
        return $topic;
    }

    /**
     * Return a `Subscription` instance from a channel name.
     *
     * If the subscription doesn't exist, the subscription is first created.
     *
     * @param string $channel
     *
     * @return \Google\Cloud\PubSub\Subscription
     */
    protected function getSubscriptionForChannel($channel)
    {
        $topic = $this->getTopicForChannel($channel);
        $clientIdentifier = $this->clientIdentifier ? $this->clientIdentifier : 'default';
        $clientIdentifier .= '.' . $channel;
        $subscription = $topic->subscription($clientIdentifier);
        if ($this->autoCreateSubscriptions && !$subscription->exists()) {
            $subscription->create();
        }
        return $subscription;
    }
}
